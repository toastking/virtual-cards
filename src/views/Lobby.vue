<template>
  <div>
    <h1 class="title is-1">Lobby</h1>
    <player-list />
    <b-button id="start-game-button" type="is-success" expanded v-on:click="startGame()"></b-button>
  </div>
</template>
<script lang="ts">
// Lobby to show players
import Vue from 'vue';
import PlayerList from '../components/PlayerList.vue';
import { mapState, mapActions } from 'vuex';
import { State } from '@/store';
export default Vue.extend({
  mounted() {
    // Get the game ID from the route and set it in the store
    const gameId = this.$route.params.gameid;
    this.$store.commit('createGameSuccess', { newGameId: gameId });

    // Bind data to firestore
    this.$store.dispatch('setupGameBinding');
    this.$store.dispatch('setupPlayerBinding');
  },
  computed: mapState({
    gameStarted(state: State) {
      return state.game.game.gameStarted;
    },
  }),
  methods: { ...mapActions(['startGame']) },
  watch: {
    gameStarted(newGameStarted) {
      if (newGameStarted === true) {
        this.$store.dispatch('routeToGame');
      }
    },
  },
  components: { 'player-list': PlayerList },
});
</script>
