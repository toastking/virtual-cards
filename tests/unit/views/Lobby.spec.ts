import { Store } from 'vuex-mock-store';
import { shallowMount, createLocalVue, mount } from '@vue/test-utils';
import Lobby from '@/views/Lobby.vue';
import VueRouter from 'vue-router';
import Buefy from 'buefy';

describe('Lobby View', () => {
  const store = new Store({
    state: { game: { game: { gameStarted: false } }, player: { players: [] } },
  });

  const localVue = createLocalVue();
  localVue.use(VueRouter);
  localVue.use(Buefy);

  const router = new VueRouter({
    routes: [
      {
        path: '/lobby/:gameid',
        name: 'lobby',
        component: Lobby,
      },
    ],
  });

  // add other mocks here so they are accessible in every component
  const mocks = {
    $store: store,
  };

  afterEach(() => {
    store.reset();
  });

  test('dispatches actions to bind the state on load', () => {
    shallowMount(Lobby, { mocks, localVue, router });
    expect(store.dispatch).toHaveBeenCalledWith('setupGameBinding');
    expect(store.dispatch).toHaveBeenCalledWith('setupPlayerBinding');
  });

  test('sets the game id based on the router', () => {
    router.push('/lobby/xyz');
    shallowMount(Lobby, { mocks, localVue, router });
    expect(store.commit).toHaveBeenCalledWith('createGameSuccess', {
      newGameId: 'xyz',
    });
  });

  test('routes to the game when the game is started ', () => {
    shallowMount(Lobby, { mocks, localVue, router });
    store.state.game.game.gameStarted = true;
    expect(store.dispatch).toHaveBeenCalledWith('routeToGame');
  });

  test('starts the game when Start Game is clicked', () => {
    const wrapper = mount(Lobby, { mocks, localVue, router });
    wrapper.find('#start-game-button').trigger('click');
    expect(store.dispatch).toHaveBeenCalledWith('startGame');
  });
});
