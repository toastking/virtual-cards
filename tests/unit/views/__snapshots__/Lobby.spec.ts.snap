// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lobby View renders the lobby 1`] = `
<section
  class="section"
>
  <h1
    class="title is-1"
  >
    Lobby
  </h1>
   
  <div
    class="container has-background-info has-text-white columns is-vcentered"
  >
    <div
      class="share-text column is-three-quarters"
    >
      Share! http://localhost/#/start/undefined
    </div>
     
    <div
      class="column"
    >
      <b-button-stub
        iconleft="copy"
        inverted="true"
        nativetype="button"
        outlined="true"
        tag="button"
        type="is-info"
      >
        Copy
      </b-button-stub>
    </div>
  </div>
   
  <div
    class="container"
  >
    <player-list-stub />
     
    <b-button-stub
      expanded="true"
      id="start-game-button"
      nativetype="button"
      tag="button"
      type="is-success is-large"
    >
      Start Game
    </b-button-stub>
  </div>
</section>
`;
